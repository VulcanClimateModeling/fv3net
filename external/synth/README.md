Regression Testing
==================

The package `synth` provides tools for generating synthetic xarray datasets for testing purposes. Compared to real data, it is much easier and cheaper to manage code for generating synthetic using software version control tools like git.

Usage
-----

This packages allows the user to build up a "schema" describing their data. A schema is a reduced description of a zarr/xarray dataset that can

1.  be easily serialized to disk and loaded again
2.  used to generate a random dataset
3.  validate an existing dataset

This package defines a set of dataclasses defining a schema. It also contains fixtures for datasets used by VCM-ML in training models, allowing for unit and regression testing of data pipelines.

The function synth.read\_schema\_from\_zarr can be used to generate a schema from an existing dataset. For example, suppose we have loaded a zarr group like this:

    import sys
    import fsspec
    import zarr
    import synth

    url = "gs://path/to/data.zarr"
    mapper = fsspec.get_mapper(url)
    group = zarr.open_group(mapper)

This data could comprise many gigabytes, so it is unweildy to manage, and use within a testing framework. To generate a condensed description, you can generate a reduced "schema" like this:

    schema = synth.read_schema_from_zarr(
        group, coords=("time", "tile", "grid_xt", "grid_yt")
    )

Next, `schema` can be serialized to json and dumped to disk like this:

    with open("schema.json", "w") as f:
        synth.dump(schema, f)

The json file includes a "version" entry to ensure that schema's generated by older versions of this code remain legible to new versions.

> **note**
>
> While old schema may be read by newer versions of the code, there is no guarantee that `generate` will return the same values, so checks of bit-for-bit reproducibilty may need to be updated.

The json file can then be checked into version control and loaded inside of a test script like this:

    with open("schema.json" , "r") as f:
        schema = synth.load(f)

A fake xarray dataset can be created:

    ds = synth.generate(schema)

The data for each chunk will be identically generated from a uniform distribution. The upper and lower bounds of this distribution are set by default in the code, but can be overrided using the `ranges` argument of `synth.generate`.

A dataset can be validated by checking its schema against a reference schema:

    with open("schema.json" , "r") as f:
        ref_schema = synth.load(f)

    zarr_schema = synth.read_schema_from_zarr(
        group, coords=("time", "tile", "grid_xt", "grid_yt")
    )

    assert zarr_schema == ref_schema

Note that the equality operator `==` checks the following:

1.  the dataset has the same coordinates in terms of names and dimension names
2.  the dataset has the same variables in terms of names, dimension names, and data shape, type, and chunks.

Note that coordinate shapes and values and dataset, coordinate, and variable attributes are not compared. The equality operator can also be used directly on coordinate and variable schema.

Pytest dataset fixtures
-----------------------

The package contains fixtures for two datasets used for ML training in fv3net: nudging tendencies, and fine-res apparent sources. The fixtures in `synth._dataset_fixtures` may be imported into a `conftest.py` file and then used in testing mappers and other functions that load or use these data:: from synth import ( dataset\_fixtures\_dir, data\_source\_name, nudging\_dataset\_path, fine\_res\_dataset\_path, data\_source\_path, grid\_dataset, )

Fixtures exist for each invididual dataset (e.g, `nudging_dataset_path`), returning its path in a temporary testing directory, and for a parametrized fixture (`data_source_path`) that will sequentially return the paths of all datasets.

Existing tools
--------------

Python has some [rich tools](https://faker.readthedocs.io/en/master/) for generating fake data, but nothing specialized to xarray/zarr.
