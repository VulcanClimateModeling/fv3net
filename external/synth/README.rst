Integration tests with synthetatic data

1. Generate schema of big zarr data with:

    python zarr_to_test_schema.py > schema.json

1. Run integration tests :
   
    pytest test_integration.py

Saves output to disk
