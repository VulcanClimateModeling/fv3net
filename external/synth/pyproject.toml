[tool.poetry]
name = "synth"
version = "0.1.0"
description = ""
authors = ["Noah D. Brenowitz <nbren12@gmail.com>"]

[tool.poetry.dependencies]
python = "^3.7"
fsspec = "^0.7.3"
xarray = "^0.15.1"
zarr = "^2.4.0"
toolz = "^0.10.0"
dask = "^2.15.0"

[tool.poetry.dev-dependencies]
pytest = "^5.2"
pytest-regtest = "^1.4.4"

[build-system]
requires = ["poetry>=0.12"]
build-backend = "poetry.masonry.api"
