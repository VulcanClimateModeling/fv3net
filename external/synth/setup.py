# -*- coding: utf-8 -*-
try:
    from setuptools import setup
except ImportError:
    from distutils.core import setup

readme = ""

setup(
    long_description=readme,
    name="synth",
    version="0.1.0",
    python_requires="==3.*,>=3.6.0",
    author="Noah D. Brenowitz",
    author_email="noahb@vulcan.com",
    packages=["synth"],
    package_dir={"": "."},
    package_data={"synth": ["_dataset_fixtures/*.json"]},
    install_requires=[
        "dask==2.*,>=2.15.0",
        "fsspec==0.*,>=0.7.3",
        "toolz==0.*,>=0.10.0",
        "xarray==0.*,>=0.15.1",
        "zarr==2.*,>=2.4.0",
    ],
    extras_require={"dev": ["pytest==5.*,>=5.2.0", "pytest-regtest==1.*,>=1.4.4"]},
)
