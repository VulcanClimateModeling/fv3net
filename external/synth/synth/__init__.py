from .core import (Array, ChunkedArray, CoordinateSchema, DatasetSchema,
                   Domain, MyEncoder, Range, VariableSchema, 
                   dumps, loads, load, dump, sample)


__version__ = '0.1.0'
