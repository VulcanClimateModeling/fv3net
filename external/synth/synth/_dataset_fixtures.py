import os
import tempfile
import numpy as np
import xarray as xr
import pytest
from distutils import dir_util

from .core import load, generate, Range


timestep1 = "20160801.001500"
timestep1_npdatetime_fmt = "2016-08-01T00:15:00"
timestep2 = "20160801.003000"
timestep2_npdatetime_fmt = "2016-08-01T00:30:00"


@pytest.fixture(scope="module")
def dataset_fixtures_dir(tmpdir_factory, request):
    """Creates a temporary directory for the contents of the
    synth datasets for use in dataset fixtures, and returns its path"""

    test_dir, _ = os.path.splitext(__file__)

    tmpdir = tmpdir_factory.mktemp("pytest_data")

    if os.path.isdir(test_dir):
        dir_util.copy_tree(test_dir, str(tmpdir))

    return tmpdir


@pytest.fixture(
    params=["one_step_tendencies", "nudging_tendencies", "fine_res_apparent_sources"]
)
def data_source_name(request):
    return request.param


@pytest.fixture(scope="module")
def one_step_dataset_path(dataset_fixtures_dir):

    with tempfile.TemporaryDirectory() as one_step_dir:
        _generate_one_step_dataset(dataset_fixtures_dir, one_step_dir)
        yield one_step_dir


def _generate_one_step_dataset(datadir, one_step_dir):

    with open(str(datadir.join("one_step.json"))) as f:
        one_step_schema = load(f)
    one_step_dataset = generate(one_step_schema)
    one_step_dataset_1 = one_step_dataset.assign_coords({"initial_time": [timestep1]})
    one_step_dataset_2 = one_step_dataset.assign_coords({"initial_time": [timestep2]})
    one_step_dataset_1.to_zarr(
        os.path.join(one_step_dir, f"{timestep1}.zarr"), consolidated=True,
    )
    one_step_dataset_2.to_zarr(
        os.path.join(one_step_dir, f"{timestep2}.zarr"), consolidated=True,
    )


@pytest.fixture(scope="module")
def nudging_dataset_path(dataset_fixtures_dir):

    with tempfile.TemporaryDirectory() as nudging_dir:
        _generate_nudging_dataset(dataset_fixtures_dir, nudging_dir)
        yield nudging_dir


def _generate_nudging_dataset(datadir, nudging_dir):

    nudging_after_dynamics_zarrpath = os.path.join(
        nudging_dir, "outdir-3h", "after_dynamics.zarr"
    )
    with open(str(datadir.join("after_dynamics.json"))) as f:
        nudging_after_dynamics_schema = load(f)
    nudging_after_dynamics_dataset = generate(
        nudging_after_dynamics_schema
    ).assign_coords(
        {
            "time": [
                np.datetime64(timestep1_npdatetime_fmt),
                np.datetime64(timestep2_npdatetime_fmt),
            ]
        }
    )
    nudging_after_dynamics_dataset.to_zarr(
        nudging_after_dynamics_zarrpath, consolidated=True
    )

    nudging_after_physics_zarrpath = os.path.join(
        nudging_dir, "outdir-3h", "after_physics.zarr"
    )
    with open(str(datadir.join("after_physics.json"))) as f:
        nudging_after_physics_schema = load(f)
    nudging_after_physics_dataset = generate(
        nudging_after_physics_schema
    ).assign_coords(
        {
            "time": [
                np.datetime64(timestep1_npdatetime_fmt),
                np.datetime64(timestep2_npdatetime_fmt),
            ]
        }
    )
    nudging_after_physics_dataset.to_zarr(
        nudging_after_physics_zarrpath, consolidated=True
    )

    nudging_tendencies_zarrpath = os.path.join(
        nudging_dir, "outdir-3h", "nudging_tendencies.zarr"
    )
    with open(str(datadir.join("nudging_tendencies.json"))) as f:
        nudging_tendencies_schema = load(f)
    nudging_tendencies_dataset = generate(nudging_tendencies_schema).assign_coords(
        {
            "time": [
                np.datetime64(timestep1_npdatetime_fmt),
                np.datetime64(timestep2_npdatetime_fmt),
            ]
        }
    )
    nudging_tendencies_dataset.to_zarr(nudging_tendencies_zarrpath, consolidated=True)


@pytest.fixture(scope="module")
def fine_res_dataset_path(dataset_fixtures_dir):

    with tempfile.TemporaryDirectory() as fine_res_dir:
        fine_res_zarrpath = _generate_fine_res_dataset(
            dataset_fixtures_dir, fine_res_dir
        )
        yield fine_res_zarrpath


def _generate_fine_res_dataset(datadir, fine_res_dir):
    """ Note that this does not follow the pattern of the other two datasets
    in that the synthetic data are not stored in the original format of the
    fine res data (tiled netcdfs), but instead as a zarr, because synth does
    not currently support generating netcdfs or splitting by tile
    """

    fine_res_zarrpath = os.path.join(fine_res_dir, "fine_res_budget.zarr")
    with open(str(datadir.join("fine_res_budget.json"))) as f:
        fine_res_budget_schema = load(f)
    fine_res_budget_dataset = generate(fine_res_budget_schema)
    fine_res_budget_dataset_1 = fine_res_budget_dataset.assign_coords(
        {"time": [timestep1]}
    )
    fine_res_budget_dataset_2 = fine_res_budget_dataset.assign_coords(
        {"time": [timestep2]}
    )
    fine_res_budget_dataset = xr.concat(
        [fine_res_budget_dataset_1, fine_res_budget_dataset_2], dim="time"
    )
    fine_res_budget_dataset.to_zarr(fine_res_zarrpath, consolidated=True)

    return fine_res_zarrpath


@pytest.fixture
def data_source_path(dataset_fixtures_dir, data_source_name):
    with tempfile.TemporaryDirectory() as data_dir:
        if data_source_name == "one_step_tendencies":
            _generate_one_step_dataset(dataset_fixtures_dir, data_dir)
            data_source_path = data_dir
        elif data_source_name == "nudging_tendencies":
            _generate_nudging_dataset(dataset_fixtures_dir, data_dir)
            data_source_path = data_dir
        elif data_source_name == "fine_res_apparent_sources":
            fine_res_zarrpath = _generate_fine_res_dataset(
                dataset_fixtures_dir, data_dir
            )
            data_source_path = fine_res_zarrpath
        else:
            raise NotImplementedError()
        yield data_source_path


@pytest.fixture
def grid_dataset(dataset_fixtures_dir):
    random = np.random.RandomState(0)
    with open(str(dataset_fixtures_dir.join("grid_schema.json"))) as f:
        grid_schema = load(f)
    grid_ranges = {"area": Range(1, 2)}
    grid = generate(grid_schema, ranges=grid_ranges).load()
    grid["land_sea_mask"][:] = random.choice(
        [0, 1, 2], size=grid["land_sea_mask"].shape
    )
    return grid
