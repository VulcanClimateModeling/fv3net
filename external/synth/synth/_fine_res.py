from typing import Sequence
from synth.schemas import load_schema
from synth.core import generate
import os


def generate_fine_res(datadir: str, times: Sequence[str]):
    """Generate a directory of fine-res data
    
    Args:
        datadir: output location
        times: list of YYYYMMDD.HHMMSS timestamps
    
    """
    schema = load_schema("fine_res_budget.json")
    dataset = generate(schema)

    for tile in range(1, 7):
        for time in times:
            path = os.path.join(datadir, f"{time}.tile{tile}.nc")
            dataset.isel(tile=tile-1).to_netcdf(path)
