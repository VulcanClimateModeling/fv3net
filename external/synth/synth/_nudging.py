import os
from typing import Sequence

import numpy as np

from synth.core import DatasetSchema, generate

from .schemas import load_schema as _load_schema


def _generate(
    directory: str,
    tendencies_schema: DatasetSchema,
    after_dynamics_schema: DatasetSchema,
    after_physics_schema: DatasetSchema,
    times: Sequence[np.datetime64],
):
    nudging_dir = directory
    nudging_after_dynamics_zarrpath = os.path.join(nudging_dir, "after_dynamics.zarr")
    nudging_after_dynamics_dataset = generate(after_dynamics_schema).assign_coords(
        {"time": times}
    )
    nudging_after_dynamics_dataset.to_zarr(
        nudging_after_dynamics_zarrpath, consolidated=True
    )

    nudging_after_physics_zarrpath = os.path.join(nudging_dir, "after_physics.zarr")
    nudging_after_physics_dataset = generate(after_physics_schema).assign_coords(
        {"time": times}
    )
    nudging_after_physics_dataset.to_zarr(
        nudging_after_physics_zarrpath, consolidated=True
    )

    nudging_tendencies_zarrpath = os.path.join(nudging_dir, "nudging_tendencies.zarr")
    nudging_tendencies_dataset = generate(tendencies_schema).assign_coords(
        {"time": times}
    )
    nudging_tendencies_dataset.to_zarr(nudging_tendencies_zarrpath, consolidated=True)


def generate_nudging(outdir: str, times: Sequence[np.datetime64]):
    _generate(
        outdir,
        after_dynamics_schema=_load_schema("after_dynamics.json"),
        after_physics_schema=_load_schema("after_physics.json"),
        tendencies_schema=_load_schema("nudging_tendencies.json"),
        times=times,
    )
