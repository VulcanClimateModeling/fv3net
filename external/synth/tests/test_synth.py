import pickle

import numpy as np
import zarr

from synth import (
    read_schema_from_zarr,
    Array,
    ChunkedArray,
    CoordinateSchema,
    DatasetSchema,
    Domain,
    MyEncoder,
    Range,
    VariableSchema,
    __version__,
    dumps,
    loads,
)


def test_version():
    assert __version__ == "0.1.0"


def test_generate_array():
    r = Range(0, 1)
    a = Array(shape=(1, 2), dtype=np.float32)
    out = a.generate(r)

    assert out.dtype == a.dtype
    assert out.shape == a.shape


def test_generate_chunked_array():
    r = Range(0, 1)
    a = ChunkedArray(shape=(10, 10), dtype=np.float32, chunks=((5, 5), (5, 5)))
    out = a.generate(r)

    assert out.dtype == a.dtype
    assert out.shape == a.shape
    assert out.chunks == a.chunks


def test_encoder_dtype():
    e = MyEncoder()
    assert e.encode(np.dtype("float32")) == '"' + np.dtype("float32").str + '"'


def test_encoder_array():
    e = MyEncoder()
    assert e.encode(np.array([1, 2, 3])) == "[1, 2, 3]"


def test_DatasetSchema_dumps():

    x = CoordinateSchema("x", ["x"], np.array([1, 2, 3]))
    a = VariableSchema(
        "a",
        ["x"],
        ChunkedArray(shape=[3], chunks=[1], dtype=np.dtype("float32")),
        domain=Range(0, 10),
    )

    ds = DatasetSchema(coords=[x], variables=[a])

    val = dumps(ds)
    assert isinstance(val, str)


def test_DatasetSchema_dumps_regression(regtest):
    x = CoordinateSchema("x", ["x"], np.array([1, 2, 3]))
    a = VariableSchema(
        "a",
        ["x"],
        ChunkedArray(shape=[3], chunks=[1], dtype=np.dtype("float32")),
        domain=Range(0, 10),
    )

    ds = DatasetSchema(coords=[x], variables=[a])

    val = dumps(ds)
    print(val, file=regtest)


def test_DatasetSchemaLoads():
    encoded_data = """
    {"coords": [{"name": "x", "dims": ["x"], "value": [1, 2, 3], "attrs": [1]}], "variables": [{"name": "a", "dims": ["x"], "array": {"shape": [3], "dtype": "<f4", "chunks": [1]}, "domain": {"min": 0, "max": 10}}]}
    """  # noqa

    ds = loads(encoded_data)
    assert isinstance(ds, DatasetSchema)

    v = ds.variables[0]
    assert isinstance(v, VariableSchema)
    assert isinstance(v.domain, Domain)

    assert ds.coords[0].attrs == [1]


def test_generate_and_pickle_integration():
    x = CoordinateSchema("x", ["x"], np.array([1, 2, 3]))
    a = VariableSchema(
        "a",
        ["x"],
        ChunkedArray(shape=[3], chunks=[1], dtype=np.dtype("float32")),
        domain=Range(0, 10),
    )

    ds = DatasetSchema(coords=[x], variables=[a])
    d = ds.generate()
    pickle.dumps(d)


def test_generate_regression(regtest):
    x = CoordinateSchema("x", ["x"], np.array([1, 2, 3]))
    a = VariableSchema(
        "a",
        ["x"],
        ChunkedArray(shape=[3], chunks=[1], dtype=np.dtype("float32")),
        domain=Range(0, 10),
    )

    ds = DatasetSchema(coords=[x], variables=[a])
    d = ds.generate()
    arr = d.a.values
    print(arr, file=regtest)


def test_cftime_generate():
    julian_time_attrs = {
        "calendar": "julian",
        "calendar_type": "JULIAN",
        "cartesian_axis": "T",
        "long_name": "time",
        "units": "seconds since 2016-08-01T00:15:00.000026",
    }

    store = {}

    group = zarr.open_group(store, mode="w")
    arr = group.zeros("time", shape=[1], dtype=np.float64)
    arr.attrs.update(julian_time_attrs)
    arr.attrs["_ARRAY_DIMENSIONS"] = ["time"]

    schema = read_schema_from_zarr(group, coords=["time"])
    ds = schema.generate()

    assert dict(ds.time.attrs) == dict(julian_time_attrs)
