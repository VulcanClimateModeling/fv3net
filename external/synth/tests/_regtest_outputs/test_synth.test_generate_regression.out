[5.488135 5.488135 5.488135]
